File: dndacademyBE/src/main/java/com/giulia/dndacademy/model/CampaignChapter.java
package com.giulia.dndacademy.model;

import jakarta.persistence.*;
import lombok.*;

@Entity
@Table(name = "campaign_chapters")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class CampaignChapter {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(length = 1000)
    private String description;

    @Column(length = 8000, nullable = false)
    private String storyText;

    @Column(nullable = false)
    private int orderIndex;

    @Column(nullable = false)
    private boolean hasCombat;

    @ManyToOne
    @JoinColumn(name = "campaign_id", nullable = false)
    private Campaign campaign;

    @ManyToOne
    @JoinColumn(name = "lesson_id")
    private Lesson lesson;

    @ManyToOne
    @JoinColumn(name = "quiz_id")
    private Quiz quiz;

    @ManyToOne
    @JoinColumn(name = "reward_badge_id")
    private Badge rewardBadge;
}

File: dndacademyfe/src/types/campaign-chapter.ts
export type CampaignChapterPlayer = {
  chapterId: number
  campaignId: number
  campaignName: string
  title: string
  description: string | null
  storyText: string
  orderIndex: number
  hasCombat: boolean
  lessonId: number | null
  lessonTitle: string | null
  quizId: number | null
  quizTitle: string | null
  rewardBadgeId: number | null
  rewardBadgeName: string | null
  unlocked: boolean
  completed: boolean
  completedAt: string | null
}

export type CampaignChapter = {
  id: number
  campaignId: number
  campaignName: string
  title: string
  description: string | null
  storyText: string
  orderIndex: number
  hasCombat: boolean
  lessonId: number | null
  lessonTitle: string | null
  quizId: number | null
  quizTitle: string | null
  rewardBadgeId: number | null
  rewardBadgeName: string | null
}

export type CreateCampaignChapterRequest = {
  title: string
  description: string
  storyText: string
  orderIndex: number
  hasCombat: boolean
  lessonId: number | null
  quizId: number | null
  rewardBadgeId: number | null
}
